src/index.js

```diff
@@ -72,30 +72,37 @@ function glamorous(comp, {rootEl, displayName, forwardProps = []} = {}) {
       }
 
       render() {
-        const {className, innerRef, ...rest} = this.props
-        const {toForward, cssOverrides} = splitProps(rest, GlamorousComponent)
+        // in this function, we're willing to sacrafice a little on
+        // readability to get better performance because it actually
+        // matters.
+        const props = this.props
+        const {toForward, cssOverrides} = splitProps(props, GlamorousComponent)
+        // freeze the theme object in dev mode
+        const theme = process.env.NODE_ENV === 'production' ?
+          this.state.theme :
+          Object.freeze(this.state.theme)
         // create className to apply
-        const mappedArgs = GlamorousComponent.styles.map(glamorRules => {
-          if (typeof glamorRules === 'function') {
-            return glamorRules(this.props, {...this.state.theme})
+        const mappedArgs = GlamorousComponent.styles.slice()
+        for (let i = mappedArgs.length; i--;) {
+          if (typeof mappedArgs[i] === 'function') {
+            mappedArgs[i] = mappedArgs[i](props, theme)
           }
-          return glamorRules
-        })
+        }
         const {
           glamorStyles: parentGlamorStyles,
           glamorlessClassName,
-        } = extractGlamorStyles(className)
+        } = extractGlamorStyles(props.className)
         const glamorClassName = css(
           ...mappedArgs,
           ...parentGlamorStyles,
           cssOverrides,
         ).toString()
-        const fullClassName = joinClasses(glamorlessClassName, glamorClassName)
```

Comment:
I figured that calling out to a function that does all the stuff `joinClasses` did was unnecessary weight when the same could be done with simple string concat and `trim`.